hbase-client/src/main/java/org/apache/hadoop/hbase/MetaTableAccessor.java

```diff
@@ -1846,6 +1846,16 @@ public static void deleteMergeQualifiers(Connection connection, final RegionInfo
       qualifiers.add(qualifier);
       delete.addColumns(getCatalogFamily(), qualifier, HConstants.LATEST_TIMESTAMP);
     }
+
+    // There will be race condition that a GCMultipleMergedRegionsProcedure is scheduled while
+    // the previous GCMultipleMergedRegionsProcedure is still going on, in this case, the second
+    // GCMultipleMergedRegionsProcedure could delete the merged region by accident!
+    if (qualifiers.isEmpty()) {
```

Comment:
Thanks Stack, where to add the lock? More context here, this issue was found with the extra frequent calling of CatalogJanitor.scan() in TestMetaFixer#testOverlap(). It finds merged regions in meta table and fires GCMultipleMergedRegionsProcedure to gc these regions. The second scan() could find the same set of merged regions in meta and keep firing GCMultipleMergedRegionsProcedures. So the first GCMultipleMergedRegionsProcedure cleans up merge qualifiers and the following one will delete the new merged region from meta table, this will result in a hole. 

If we maintain a locks in Catalog Janitor, then it needs to wait for GCMultipleMergedRegionsProcedures and unlocks the region, this will make scan() very slow. What do you think?

Reply:
Thanks Stack, I will take a look at examples and come back.

Reply:
@saintstack I checked the code and went through what you said above. My understanding is that the lock prevents two procedures working on the same region at the same time. In this specific case, it is multiple MergeTableRegionsProcedures are submitted to clean up the merge states, so there are already locks in place. These multiple MergeTableRegionsProcedures are executed sequentially, because it does not check if there are mergeXXX qualifiers to delete, the second one could delete the child region.  To make MergeTableRegionsProcedures idempotent, we need this fix here. What do you think? There is another jira HBASE-24250 created by Andrey, where multiple MergeTableRegionsProcedures are observed in running clusters.

Reply:
Let me create a new issue to track the GCMulitpleMergedRegionsProcedure non-idempotent issue and come up with an unitest to show the issue.

Reply:
ok. sounds good. Let me look at the patch again.

Reply:
Thanks @saintstack . I create HBASE-24256 to track GCMulitpleMergedRegionsProcedure issue, will drop the fix code in this patch and upload a new patch. In HBASE-24256, I will add an unitest to show the issue.